Add dark mode

hlfan opened the pull request:
Fixes #8214.
This builds on the original work from @AntonKhorev to make it able to sync with the OSM preference. See https://github.com/openstreetmap/openstreetmap-website/pull/6323 for that.

Making the theme respond anytime results in variables like `--bg-filter` and `--bg-blend-mode` to not need to watch for media query changes in JS.
Other styling changes away from JS have already been made in #11292.

AntonKhorev committed:
WIP

hlfan committed:
WIP

hlfan committed:
Allow editor theme to be overwritten

k-yle commented on css/10_colors.css:
```diff
@@ -0,0 +1,177 @@
+}
+
+@media (prefers-color-scheme: dark) {
+    .ideditor:not([data-theme="light"]) {
```
if we're going to use [data attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/How_to/Use_data_attributes), should they be documented in [./API.md](https://github.com/openstreetmap/id/blob/develop/API.md)? That file is currently a collection of every way that consumers can customise iD.

(for context) I see that the OSM website [already adds 2 `data-*` attributes to the iD container](https://github.com/hlfan/openstreetmap-website/blob/4dff9d60cec071b1eb6d58b7a6d19c5d887f305f/app/views/site/id.html.erb#L17-L18), but it seems like iD doesn't actually use these (?)

hlfan replied:
These element datasets are used for the core context initialization, so setting it from there would be aligned better with the other options.

But since this PR is already quite large, I'd like to keep this CSS-only and have the JS api integration done separately.
I could see `.theme-light` or `.theme-dark` be set with
```js
iD.coreContext().theme(container.dataset.theme).init()
```
However, I don't feel comfortable with making changes there yet.

k-yle replied:
Until now, the `data-` attributes were only used by the OSM-website to pass data from [one of their files](https://github.com/hlfan/openstreetmap-website/blob/4dff9d60cec071b1eb6d58b7a6d19c5d887f305f/app/views/site/id.html.erb#L18) to [another internal file](https://github.com/hlfan/openstreetmap-website/blob/4dff9d60cec071b1eb6d58b7a6d19c5d887f305f/app/assets/javascripts/id.js#L26). iD has no code related to these `data-` attributes. So this PR is introducing root `data-` attributes to the iD codebase for the first time.

---

but if your intention is to introduce some sort of `.theme()` method to `coreContext` which updates the `data-theme`, then there's no issue, because no one would directly interact with the `data-` attributes, so we don't need to document it.

hlfan replied:
Do you think CSS classes would be better suited for this than data attributes?

k-yle replied:
if `coreContext` can be used to set the theme, then imo it doesn't matter what method is used internally. maybe @tyrasd has further thoughts

hlfan replied:
```diff
diff --git a/modules/core/context.js b/modules/core/context.js
index 3219ff60e..d3be29955 100644
--- a/modules/core/context.js
+++ b/modules/core/context.js
@@ -419,12 +419,32 @@ export function coreContext() {
   };
 
 
+  /* Theme */
+  const availableThemes = ['dark', 'light'];
+  let _theme;
+  if (availableThemes.includes(context.initialHashParams.theme)) {
+    _theme = context.initialHashParams.theme;
+  }
+  function applyTheme() {
+    availableThemes.forEach(theme =>
+      context.container().classed(`theme-${theme}`, theme === _theme)
+    );
+  };
+  context.theme = function(val) {
+    if (!arguments.length) return _theme;
+    if (availableThemes.includes(val)) _theme = val;
+    else _theme = undefined;
+    applyTheme();
+    return context;
+  };
+
   /* Container */
   let _container = d3_select(null);
   context.container = function(val) {
     if (!arguments.length) return _container;
     _container = val;
     _container.classed('ideditor', true);
+    applyTheme();
     return context;
   };
   context.containerNode = function(val) {
```

That's what I've bodged together for the context side so far. Not including this in the PR so it can be squash-merged separately.

k-yle replied:
cool, since it'll be controllable through coreContext I think we can resolve this thread, whether we use classes or `data-` attributes internally won't matter

k-yle commented on css/80_app.css:
```diff
@@ -2315,6 +2317,7 @@ input.date-selector {
 
 .form-field-input-restrictions .restriction-container {
     position: relative;
+    color: #333;
```
a few lines below here is `.restriction-help`, which has the same bg/fg colours in light and dark mode, making it noticably brighter than everything else

just checking if that's intentional?

<img height="300" alt="image" src="https://github.com/user-attachments/assets/bb72bc37-138f-49d9-8573-312e7804e128" />

hlfan replied:
I noticed it too on adopting the original work, but I left it for now since it didn't seem as blinding. However, I can take another look at it.

k-yle reviewed (commented):
I used this PR for a few edits and i think it looks really nice :) just my personal opinion, i don't know much about colour contrast guidelines in dark mode

one question about the design below:

hlfan commented:
I've been using https://pr-11308--ideditor.netlify.app/dist/iD.css for regular mapping too for a few weeks and couldn't find other light mode remains.

That said, I also don't know much about colour contrast guidelines in dark mode. But just from the hwb() way I converted the colors, at least the ones I added _should_ yield a higher contrast.

tyrasd committed:
very slightly adjust link/button colors in dark mode

tyrasd committed:
fix color scheme of turn restriction editor help dialog/overlay

also slightly tweaks styling (rounded corners), close button to better match the style of related UI elements

tyrasd committed:
drop unused commented out code and respective styles

tyrasd committed:
fix colors of error state of background-offset widget input

tyrasd committed:
slightly tweak secordary border color in dark mode

and fix color of inactive UI elements in hover state

tyrasd committed:
fix background of "more" text for community list in dark mode

tyrasd committed:
simplify keyboard shortcut help dialog styling

* use regular text color for unselected tabs
* use regular text color for keys description in table

tyrasd committed:
tweak color of "dark" tooltips

these are used on the left sidebar to preserve background against the background
* in light mode: use white text on gray background (instead of gray on black, which was quite harsh)
* in dark mode: use gray text on black background

tyrasd committed:
use defined colors in a few more places

tyrasd reviewed (approved):
I found a few niche additional UI elements that still needed special styling for dark mode (fixed already; see added commits to the branch):

<img width="300" src="https://github.com/user-attachments/assets/5bd42e93-2287-4cc5-9d38-ea4123bcd463" />

<img width="300" height="484" alt="image" src="https://github.com/user-attachments/assets/7bdde8bc-709b-4117-b371-761df0cafa22" />

<img width="300" src="https://github.com/user-attachments/assets/c9db94c5-e3dc-43f2-86a0-49ae776cadcc" />

And I took the opportunity to slightly clean up the styling in a couple of places (e.g. removing some unused styles, reducing hardcoded colors in `80_app.css`, etc., see the remaining added commits). I hope that's fine.

tyrasd committed:
add to changelog

hlfan commented:
> I hope that's fine.

Totally!
The `.field-help-nav-item:hover` background should probably be adjusted as well.

Also [Anton Khorev](https://github.com/AntonKhorev) should be credited too.

And as for the `coreContext` integration, please take a look at https://github.com/openstreetmap/iD/pull/11308#discussion_r2289893294.

Nekzuris commented:
Thank you for this!

But I think it lacks contrast, the background is too light.
[This](https://userstyles.world/style/15596/openstreetmap-dark-theme) is imo better.

tyrasd commented:
> the background is too light

My understanding is that this was based on the colors of the dark theme of the osm website, which is also not using a pitch-black background. Feel free to propose tweaks to the theme, though.

hlfan commented:
Pitch black on an HDR monitor looks worse, and while I do think the backgrounds could be a touch darker, having the colors lining up with the OSM website was a thing I left as is from the draft of @AntonKhorev.

hlfan commented:
Another file that should be updated is ACCESSIBILITY.md, as it mentions dark mode at its end.
